components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java

```diff
@@ -0,0 +1,481 @@
+ */
+
+package org.wso2.carbon.apimgt.gateway.handlers.security.jwt;
+
+import io.swagger.v3.oas.models.OpenAPI;
+import org.apache.axis2.Constants;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.synapse.MessageContext;
+import org.apache.synapse.core.axis2.Axis2MessageContext;
+import org.apache.synapse.rest.RESTConstants;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.wso2.carbon.apimgt.gateway.MethodStats;
+import org.wso2.carbon.apimgt.gateway.handlers.security.APISecurityConstants;
+import org.wso2.carbon.apimgt.gateway.handlers.security.APISecurityException;
+import org.wso2.carbon.apimgt.gateway.handlers.security.AuthenticationContext;
+import org.wso2.carbon.apimgt.gateway.internal.ServiceReferenceHolder;
+import org.wso2.carbon.apimgt.gateway.utils.OpenAPIUtils;
+import org.wso2.carbon.apimgt.impl.APIConstants;
+import org.wso2.carbon.apimgt.impl.APIManagerConfiguration;
+import org.wso2.carbon.base.MultitenantConstants;
+import org.wso2.carbon.context.PrivilegedCarbonContext;
+import org.wso2.carbon.identity.oauth.config.OAuthServerConfiguration;
+
+import javax.cache.Cache;
+import javax.cache.Caching;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.PublicKey;
+import java.security.Signature;
+import java.security.SignatureException;
+import java.security.NoSuchAlgorithmException;
+import java.security.InvalidKeyException;
+import java.security.cert.Certificate;
+import java.util.Base64;
+
+/**
+ * A Validator class to validate JWT tokens in an API request.
+ */
+public class JWTValidator {
+
+    private static final Log log = LogFactory.getLog(JWTValidator.class);
+
+    private String apiLevelPolicy;
+    private boolean isGatewayTokenCacheEnabled;
+
+    public JWTValidator(String apiLevelPolicy) {
+        this.apiLevelPolicy = apiLevelPolicy;
+        this.isGatewayTokenCacheEnabled = isGatewayTokenCacheEnabled();
+    }
+
+    /**
+     * Authenticates the given request with a JWT token to see if an API consumer is allowed to access
+     * a particular API or not.
+     *
+     * @param jwtToken             The JWT token sent with the API request
+     * @param synCtx               The message to be authenticated
+     * @param openAPI              The OpenAPI object of the invoked API
+     * @return an AuthenticationContext object which contains the authentication information
+     * @throws APISecurityException in case of authentication failure
+     */
+    @MethodStats
+    public AuthenticationContext authenticate(String jwtToken, MessageContext synCtx, OpenAPI openAPI)
+            throws APISecurityException {
+
+        String[] splitToken = jwtToken.split("\\.");
+
+        JSONObject payload = null;
+        boolean isVerified = false;
+
+        String tokenSignature = splitToken[2];
```

Comment:
possible  array index out of bounds if someone sends a invalid jwt?

Reply:
Invalid token length is handled in the OAuthAuthenticator before hitting this code